numba_cuda/numba/cuda/typing/typeof.py

```diff
@@ -55,6 +55,14 @@ def typeof_impl(val, c):
     if tp is not None:
         return tp
 
+    # Check for __cuda_array_interface__ objects (third-party device arrays)
+
+    # Numba's own DeviceNDArray is handled above via _numba_type_.
+    if hasattr(val, "__cuda_array_interface__"):
```

Comment:
Same review as above regarding `getattr`. Can you make `_typeof_cuda_array_interface` take the interface `dict` and then just access the attribute once?